packages/flutter/lib/src/widgets/routes.dart

```diff
@@ -428,6 +428,137 @@ abstract class TransitionRoute<T> extends OverlayRoute<T> {
+  ///     begin: Offset.zero,
+  ///     end: const Offset(0.0, 1.0),
+  ///   ).animate(secondaryAnimation),
+  ///   child: child,
+  /// );
+  /// ```
+  ///
+  /// In practice this method is used pretty rarely.
+  ///
+  /// The arguments to this method are as follows:
+  ///
+  ///  * `context`: The context in which the route is being built.
+  ///  * [secondaryAnimation]: When the Navigator pushes this route
+  ///    on the top of its stack, the previous topmost route's [secondaryAnimation]
+  ///    runs from 0.0 to 1.0. When the [Navigator] pops the topmost route, the
+  ///    [secondaryAnimation] for the route below it runs from 1.0 to 0.0.
+  ///  * `child`, the page contents from the previous route by
+  ///     previous route's [buildPage].
+  Widget buildSecondaryAnimationTransitionForPreviousRoute(
+    BuildContext context,
+    Animation<double> secondaryAnimation,
+    Widget child,
+  ) {
+    return child;
+  }
+
+  /// Returns true if the [previousRoute] bottom route should delegate the
+  /// `secondaryAnimation` transition when this route is pushed on top of it
+  /// or when then this route is popped off of it.
+  ///
+  /// Subclasses can override this method to restrict the set of routes they
+  /// need to coordinate transitions with.
+  ///
+  /// If true, and `previousRoute.canTransitionTo()` is true, then the route
+  /// [previousRoute] will delegate its `secondaryAnimation` transition to
+  /// this route's [buildSecondaryAnimationTransitionForPreviousRoute] method
+  ///
+  /// If false, [previousRoute]'s `secondaryAnimation` transition will be
+  /// handled by default by [buildTransitions]
+  ///
+  /// Returns false by default.
+  ///
+  /// See also:
+  ///  * [canTransitionTo], which must be true for [previousRoute] for the
+  ///    [getSecondaryAnimationTransitionForPreviousRoute] `secondaryAnimation` to run.
+  ///
+  ///  * [buildSecondaryAnimationTransitionForPreviousRoute], to define
+  ///    [previousRoute]'s `secondaryAnimation` transition when this is true
+  bool handleSecondaryAnimationTransitionForPreviousRoute(Route<dynamic> previousRoute) => false;
+
+
+  @override
+  void didChangeNext(Route<dynamic>? nextRoute) {
+    if (nextRoute is DelegatedTransitionsRoute) {
+      _nextRoute = nextRoute;
+    }
+    super.didChangeNext(nextRoute);
+  }
+
+  Widget _buildProxyTransitions(
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+    Widget child,
+  ) {
+    // We remove the _nextRoute when has completed animating
+    if (secondaryAnimation.status == AnimationStatus.dismissed)
+        _nextRoute = null;
+
+    if (_nextRoute != null && _nextRoute!.handleSecondaryAnimationTransitionForPreviousRoute(this)) {
+      assert(!_nextRoute!._transitionCompleter.isCompleted, 'Cannot reuse a ${_nextRoute!.runtimeType} after disposing it.');
+      final ProxyAnimation proxySecondaryAnimation = ProxyAnimation(kAlwaysDismissedAnimation);
```

Comment:
Proxying the animation here will mean that `ModalRoute.buildPage` will still see the original secondaryAnimation, right? Does that cause any problems? It definitely seems confusing that your buildTransition and buildPage see different secondaryAnimations...

Reply:
Yes. `ModalRoute.buildPage` doesn't expect to use secondaryAnimation for a translation animation. So I think it should be the original secondaryAnimation that provide feedback on the progress of the translation, otherwise it would be always dismissed.

 Taking a look at the framework it is not used anywhere. Only the animation param is used inside _SearchPageRoute.buildPage, to request focus when the page completes the transition.